### DISCUSSION TITLE
Does it make sense to round-trip ImageSpec metadata when saving?

### DISCUSSION THREAD

**Discussion by jessey-git** [OpenImageIO/oiio#3693] [General]
Imagine an app that first loads a particular image using OIIO. As part of loading it dutifully copies the loaded spec's extra attributes for display/informational purposes. It'll copy over things like "DateTime", "planarconfig", "tiff:Compression" and "compression" etc.

Now the user of this app decides to save this file back out. Maybe in the same format, maybe in a completely different format.

It seems like an app should not carry these prior saved attributes during the save regardless of destination format. Are there any situations where round-tripping should be done?

**Comment by etheory** [comment]
Hi all,

I can't think of a situation where round tripping wouldn't be desirable.

If a file was saved with data, one assumes it is for a reason, and it's
likely that within a VFX pipeline, for instance, that round tripping of
data is essential to ensure efficient automation for tools across the
pipeline.

Saving a file and getting different metadata could cause issues for
pipelines that use such metadata to make decisions up and down a pipeline.

At the very least, if the weight of public opinion is against metadata, an
option should definitely be provided to allow the user to keep it.

On Tue, 29 Nov 2022 at 15:17, Jesse Y ***@***.***> wrote:

> Imagine an app that first loads a particular image using OIIO. As part of
> loading it dutifully copies the loaded spec's extra attributes for
> display/informational purposes. It'll copy over things like "DateTime",
> "planarconfig", "tiff:Compression" and "compression" etc.
>
> Now the user of this app decides to save this file back out. Maybe in the
> same format, maybe in a completely different format.
>
> It seems like an app should not carry these prior saved attributes during
> the save regardless of destination format. Are there any situations where
> round-tripping should be done?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenImageIO/oiio/discussions/3693>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKWGRMA7OGOITBAQCDD7RTWKV7V3ANCNFSM6AAAAAASOBBJCE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Luke Emrose
aka *evolutionary theory*
www.evolutionarytheory.com
www.soundcloud.com/evolutionarytheory
www.reverbnation.com/evolutionarytheory
http://www.facebook.com/pages/evolutionary-theory/110717958952413

**Reply by jessey-git** [reply]
Hmm, perhaps, but consider the following.
Does it make sense to write out that `tiff:Compression` attribute when saving to a .EXR file?  Is it safe to leave the `compression` value of "rle" coming from a TARGA and pass that to the TIFF output with that parameter set?

Do you filter these cases out or sanitize them? How do we know which to sanitize or remove?

**Reply by lgritz** [reply]
Aha, but actually it already considers this!

https://github.com/OpenImageIO/oiio/blob/master/src/openexr.imageio/exroutput.cpp#L982

For "namespaced" metadata (i.e., in the form `namespace:attribname`) where the namespace IS THE NAME OF AN IMAGE FILE FORMAT, the OpenEXR writer will not save it to the output file.

The convention for image readers is that they prefix metadata with the format name if it's really only meaningful to that one format.

**Comment by lgritz** [comment]
In general, we've always tried to propagate the metadata from input to output. Though in practice, few file formats allow arbitrarily named metadata, so in general this won't really propagate anything other than the handful of "canonical" metadata that it tries really hard to understand, for example it knows that the "ImageDescription" field of TIFF should carry through to the "comments" block of a JPEG. 

OpenEXR is one of the few formats (maybe the only one?) that supports truly arbitrary name/value metadata that isn't defined by the spec or some extension like Exif, so it's the only one that is in any real danger of "accidentally" passing on something that really shouldn't be propagated between files of differing formats. And for that case, we do (as I said in the earlier comment) at least suppress the propagation of metadata that is prefixed by the name of any other known image file format, which are assumed to be items that it would only make sense to propagate to output files of the same file format as the input (for example, the numerical code of the compression type, or something like "tiff:RowsPerStrip" which is extremely specific to the way TIFF files are organized).

As another exception to propagation, some ImageSpec metadata items that are prefixed with "oiio:..." are understood to be hints to the writer, not real metadata to be saved in the file, and a few that start with "openexr:..." are also hints, or have some special meaning only to the exr writer.

**Reply by etheory** [reply]
I almost feel like there should be a specific distinction between "format metadata" and "general metadata", unless that's what is being referred to be "ImageSpec metadata", I thought there was only one type in each format? I guess that's up to the formats themselves if such a distinction is made. It'd be useful if it was, though.

**Comment by jessey-git** [comment]
It seems that while OpenEXR filters the values, some other formats do not?  I have a .png here with .tiff metadata :) ... and I see that there's a PR with some explanation for the "DateTime" field.

Ultimately though, the confusion really started when "compression" and "oiio:UnassociatedAlpha" values flowed between formats while I was debugging. The code was going to re-set those fields to something appropriate later on but those still worry me.

**Comment by lgritz** [comment]
I'm not even sure what you mean by "png with tiff metadata". Can you give an example?

"compression" does propagate, by design. The assumption is that if an input file used compression method that is supported by the output file format, then all other things being equal, the same compression should be used for output. All of the formats, if given a compression for output that they don't directly support, will ignore it and use their "default."

"oiio:UnassociatedAlpha" also is probably something you don't want to drop on the floor. Coming from the input, it indicates that the data is NOT premultiplied, and you definitely want to know that when you output, no?

Maybe I just need concrete examples. Like an example input file, a specific `oiiotool input ...stuff... -o output` command (or something else we can easily try on our end) that yields an output file that contains metadata from the input you think should not be propagated?

**Comment by jessey-git** [comment]
Sure, here's a png with the "tiff:" metadata which can be displayed with `iinfo -v test.png` [1] 
![test](https://user-images.githubusercontent.com/7989986/204647242-6c456e6e-dd38-4cc7-a864-07f259ba7256.png)

Here's a rundown of the steps that ended up producing the above file. The app in question basically does the following:
- TIFF load: We use OIIO to load in the image data into our own, managed, pixel buffers and we'll copy its `extra_attribs` into our own list of metadata strings
- User stuff: When the user decides they are finished with the image, they can then save it out in whatever manner they like; different format, different CM settings, different compression etc
- PNG save: For the save, we'll create a new ImageSpec and part of that process is inserting the list of metadata strings back into that new object (for round-trip purposes)

[1] Example
```
.\test.png :  128 x  128, 4 channel, uint16 png
    channel list: R, G, B, A
    DateTime: "2022:11:29 00:32:55"
    Orientation: "1"
    ResolutionUnit: "inch"
    XResolution: 2.54
    YResolution: 2.54
    oiio:BitsPerSample: "16"
    oiio:ColorSpace: "sRGB"
    tiff:Compression: "8"
    tiff:PhotometricInterpretation: "2"
    tiff:PlanarConfiguration: "1"
    tiff:RowsPerStrip: "32"
```

**Comment by jessey-git** [comment]
Here's the relevant code which can result in the above .png issue. Strangely using oiiotool directly doesn't do this.  I'm not sure what is stripping out those "tiff:" attributes in that case, but using the code itself doesn't seem to do it.

```
    // Read in the TIFF
    auto in = OIIO::ImageInput::open("t:/test.tif");

    const OIIO::ImageSpec& in_spec = in->spec();
    int xres = in_spec.width;
    int yres = in_spec.height;
    int channels = in_spec.nchannels;

    std::vector<unsigned char> pixels(xres * yres * channels);
    in->read_image(OIIO::TypeDesc::UINT8, &pixels[0]);
    in->close();

    // Time passes... The user wants to save as a PNG now
    // Create a new out_spec and copy over the extra_attribs
    // 
    // This is slightly different than what happens in the real code - we no longer have the in_spec there
    // but we do have the name/value pairs of the extra_attribs that we saved. The result is the
    // same in either case --> the "tiff:" attributes make their way out into the file

    OIIO::ImageSpec out_spec(xres, yres, channels, OIIO::TypeDesc::UINT8);
    OIIO::ParamValueList metadata = in_spec.extra_attribs;
    for (auto& attrib : metadata)
    {
        // Manual filtering would yield a "proper" result it seems (?)
        //
        // auto& name = attrib.name();
        // if (const char* colon = strchr(name.c_str(), ':')) {
        //     std::string prefix(name.c_str(), colon);
        //     OIIO::Strutil::to_lower(prefix);
        //     if (prefix == "oiio" || (prefix != "png" && OIIO::is_imageio_format_name(prefix)))
        //         continue; // SKIP!
        // }

        out_spec.attribute(attrib.name(), attrib.get_string());
    }

    auto out = OIIO::ImageOutput::create("t:/test.png");
    out->open("t:/test.png", out_spec);
    out->write_image(OIIO::TypeDesc::UINT8, pixels.data());
    out->close();
```
